llvm/include/llvm/Transforms/Utils/SYCLUtils.h

```diff
@@ -0,0 +1,137 @@
+  traverseCallgraphUp(F, CallGraphNodeAction(ActionF), Visited,
+                      ErrorOnNonCallUse, functionFilter);
+}
+
+template <class CallGraphNodeActionF>
+void traverseCallgraphUp(
+    Function *F, CallGraphNodeActionF ActionF, bool ErrorOnNonCallUse = true,
+    const CallGraphFunctionFilter &functionFilter =
+        [](const Instruction *, const Function *) { return true; }) {
+  SmallPtrSet<Function *, 32> Visited;
+  traverseCallgraphUp(F, CallGraphNodeAction(ActionF), Visited,
+                      ErrorOnNonCallUse, functionFilter);
+}
+
+/// Tells if this value is a bit cast or address space cast.
+bool isCast(const Value *V);
+
+/// Tells if this value is a GEP instructions with all zero indices.
+bool isZeroGEP(const Value *V);
+
+/// Climbs up the use-def chain of given value until a value which is not a
+/// bit cast or address space cast is met.
+const Value *stripCasts(const Value *V);
+Value *stripCasts(Value *V);
+
+/// Climbs up the use-def chain of given value until a value is met which is
+/// neither of:
+/// - bit cast
+/// - address space cast
+/// - GEP instruction with all zero indices
+const Value *stripCastsAndZeroGEPs(const Value *V);
+Value *stripCastsAndZeroGEPs(Value *V);
+
+/// Collects uses of given value "looking through" casts. I.e. if a use is a
+/// cast (chain), then uses of the result of the cast (chain) are collected.
+void collectUsesLookThroughCasts(const Value *V,
+                                 SmallPtrSetImpl<const Use *> &Uses);
+
+/// Collects uses of given pointer-typed value "looking through" casts and GEPs
+/// with all zero indices - those pointer transformation instructions which
+/// don't change pointed-to value. E.g. if a use is a cast (chain), then uses of
+/// the result of the cast (chain) are collected.
+void collectUsesLookThroughCastsAndZeroGEPs(const Value *V,
+                                            SmallPtrSetImpl<const Use *> &Uses);
+
+void collectUsesLookThroughCasts(const Value *V,
+                                 SmallPtrSetImpl<const Use *> &Uses);
+
+void collectUsesLookThroughCastsAndZeroGEPs(const Value *V,
+                                            SmallPtrSetImpl<const Use *> &Uses);
+
+bool collectPossibleStoredVals(
+    Value *Addr, SmallPtrSetImpl<Value *> &Vals,
+    std::function<bool(const CallInst *)> EscapesIfAddrIsArgOf =
+        [](const CallInst *) { return true; });
+
+inline bool isSYCLExternalFunction(const Function *F) {
+  return F->hasFnAttribute(ATTR_SYCL_MODULE_ID);
+}
+
+/// Removes the global variable "llvm.used" and returns true on success.
+/// "llvm.used" is a global constant array containing references to kernels
+/// available in the module and callable from host code. The elements of
+/// the array are ConstantExpr bitcast to i8*.
+/// The variable must be removed as it is a) has done the job to the moment
+/// of this function call and b) the references to the kernels callable from
+/// host must not have users.
```

Comment:
a) we need to explain what was the variable up to the moment of this function call. Will it even be used for the same purpose in upstream?
b) I don't understand the requirement here, do we have a more detailed documentation?

Reply:
The function is removed.